theano/tensor/opt.py

```diff
@@ -4121,16 +4140,33 @@ def local_reshape_lift(node):
             isinstance(node.inputs[0].owner.op, T.Elemwise) and
             len(node.inputs[0].owner.inputs) == 1):
         r = node.op(node.inputs[0].owner.inputs[0], node.inputs[1])
+        # Copy stacktrace from previous Reshape op, as an error in new
+        # Reshape op could only have been caused by old one.
+        copy_stack_trace(node.outputs, r)
+
         e = node.inputs[0].owner.op(r)
+        # Copy stacktrace from both previous Reshape and UnaryElemwise op
+        # because an error in new cg could have been caused by either ops.
+        copy_stack_trace(node.outputs + node.inputs, e)
+
         # In rare case the original broadcast was (False, True), but
         # the new one is (False, False). So don't crash in that case.
         if e.type != node.outputs[0].type:
-            e = T.patternbroadcast(e, node.outputs[0].broadcastable)
-        return [e]
+            re = T.patternbroadcast(e, node.outputs[0].broadcastable)
+
+            # We assume that the broadcast op cannot fail. Thus, if the
+            # graph fails it must be due to previous UnaryElemwise op, and
+            # therefore we must copy its stacktrace over.
```

Comment:
If the broadcasting operation raises an exception, that is usually due to the fact that a dimension that should be broadcastable does not actually have length 1. In that case, the responsible Op will likely be the reshape, not the elemwise.
This is just a comment on your comment, though, since the code ends up copying both original traces, which I think is OK.